### PR TITLE
[core] Ref counting for returning object IDs created by a different process

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

This is a follow-up to #6945 that adds support for the case when an ObjectID is created by one process, and another process (a borrower) is given a Python reference to the ObjectID. Then, after executing a task, the borrower stores the ObjectID inside a task return value. In this case, we must account for the process that owns the task return value in the inner object's ref count, which may be different from the borrower. This is done by:
1. The borrower stores the address of the task's caller. This is sent to the owner of the inner object ID once the borrower's own reference goes out of scope.
2. The owner of the inner ObjectID notifies the task's caller that its return value contains the ObjectID, in `WaitForRefRemovedRequest`.

We need to notify the task's caller about the nested ID in case 2 so that the task's caller knows about the nested ID *before* it processes the `WaitForRefRemovedRequest` from the owner. This is also why we can't just attach the information about the nested ID on the borrower's `PushTaskReply`; this may be received after the the task's caller hears from the owner.

This PR also refactors some of the reference counter methods to make the functionality clearer. Specifically:
1. Splits `UpdateSubmittedTaskReferences` into `UpdateSubmittedTaskReferences` and `UpdateFinishedTaskReferences`.
2. Calls `AddNestedObjectIds` when necessary, to indicate that some object that the process has a reference to also contains nested object IDs.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

File: src/ray/core_worker/reference_count_test.cc [hunk 1]
```diff
@@ -1254,43 +1246,421 @@ TEST(DistributedReferenceCountTest, TestReturnObjectIdBorrowChain) {
...
-  ASSERT_FALSE(borrower_rc.HasReference(inner_id));
-  ASSERT_FALSE(owner_rc.HasReference(inner_id));
+  borrower->rc_.RemoveLocalReference(inner_id, nullptr);
+  ASSERT_FALSE(borrower->rc_.HasReference(return_id));
+  ASSERT_FALSE(borrower->rc_.HasReference(inner_id));
+  ASSERT_FALSE(owner->rc_.HasReference(inner_id));
+}
+
+// We submit a task and submit another task that depends on the return ID. The
+// submitted task returns an object ID, which will get borrowed by the second
```
Comment (edoakes) [inline src/ray/core_worker/reference_count_test.cc]:
```suggestion
// first submitted task returns an object ID, which will get borrowed by the second
```

File: src/ray/core_worker/reference_count_test.cc [hunk 2]
```diff
@@ -1254,43 +1246,421 @@ TEST(DistributedReferenceCountTest, TestReturnObjectIdBorrowChain) {
...
+  // The owner should still have the object ID in scope because it hasn't heard
+  // from borrower yet.
+  ASSERT_TRUE(owner->rc_.HasReference(inner_id));
+
+  ASSERT_TRUE(borrower->FlushBorrowerCallbacks());
+  ASSERT_FALSE(owner->rc_.HasReference(inner_id));
+}
+
+// We submit a task and submit another task that depends on the return ID. The
+// submitted task returns an object ID, which will get borrowed by the second
```
Comment (edoakes) [inline src/ray/core_worker/reference_count_test.cc]:
```suggestion
// first submitted task returns an object ID, which will get borrowed by the second
```

File: src/ray/core_worker/core_worker.cc [hunk 3]
```diff
@@ -167,6 +160,15 @@ CoreWorker::CoreWorker(const WorkerType worker_type, const Language language,
   rpc_address_.set_raylet_id(local_raylet_id.Binary());
   rpc_address_.set_worker_id(worker_context_.GetWorkerID().Binary());
 
+  reference_counter_ = std::make_shared<ReferenceCounter>(
+      rpc_address_,
+      /*distributed_ref_counting_enabled=*/
```
Comment (edoakes) [inline src/ray/core_worker/core_worker.cc]: don't think this is necessary

**Review by edoakes — APPROVED** [review]
LGTM - the refactoring is great. I assume you also tested in python?

File: src/ray/protobuf/core_worker.proto [hunk 4]
```diff
@@ -195,12 +195,19 @@ message ObjectReferenceCount {
   bool has_local_ref = 2;
   // Any other borrowers that the worker created (by passing the ID on to them).
   repeated Address borrowers = 3;
+  // The borrower may have returned the object ID nested inside the return
+  // value of a task that it executed. This list contains all task returns that
+  // were owned by a process other than the borrower. Then, the process that
+  // owns the outer object is also a borrower for as long as it has the task
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply src/ray/protobuf/core_worker.proto]: Yes.

File: src/ray/core_worker/task_manager.cc [hunk 5]
```diff
@@ -217,7 +211,11 @@ void TaskManager::RemoveFinishedTaskReferences(
                                  inlined_ids.end());
     }
   }
-  RemoveSubmittedTaskReferences(plasma_dependencies, borrower_addr, borrowed_refs);
+
+  std::vector<ObjectID> deleted;
+  reference_counter_->UpdateFinishedTaskReferences(plasma_dependencies, borrower_addr,
+                                                   borrowed_refs, &deleted);
+  in_memory_store_->Delete(deleted);
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply src/ray/core_worker/task_manager.cc]: Hmm, I don't really have a preference, either seems fine.

**Comment by stephanie-wang** [comment]
> LGTM - the refactoring is great. I assume you also tested in python?

Yeah, seems to pass the existing tests at least.

I also added a new Python test for the scenario that should be covered with this PR. Not sure if it will pass on Travis given some of the issues we saw earlier with the other tests, but we can try it and see.

**Pull request merged**